src/parser/jobs/ast/modules/ArcanaTracking.tsx

```diff
@@ -0,0 +1,509 @@
+			this.cardStateLog[0].sleeveState = sleeveDrawLog.length
+		}
+
+		return events
+
+	}
+
+	public get getCardLogs() {
+		if (this.cardStateLog.length === 0) {
+			return []
+		}
+		return this.cardStateLog
+	}
+
+	/**
+	 * @param {number} timestamp - desired timestamp to get the state. Defaults to pull state.
+	 * @returns {CardState} - object containing the card state and last event
+	 */
+	public getCardState(timestamp = this.parser.fight.start_time): CardState | undefined {
+		const stateItem = this.cardStateLog.find(artifact => artifact.lastEvent && timestamp > artifact.lastEvent.timestamp)
+		return stateItem
+	}
+
+	/**
+	 * Adds the Arcana Buff to _prepullArcanas if it was a precast status
+	 */
+	private onPrepullArcana(event: BuffEvent) {
+		if (event.timestamp !== this.parser.fight.start_time -1) {
+			return
+		}
+		this.prepullArcanas.push(event)
+	}
+
+	/**
+	 * Determine exactly when they casted their prepull arcana
+	 * If they had overwritten this buff, it will falsly pull back the timestamp of their prepull cast, but since we are guessing, it may as well be the same.
+	 */
+	private offPrepullArcana(event: BuffEvent) {
+		if (event.timestamp >= this.parser.fight.start_time + (STATUSES.THE_BALANCE.duration * 1000)) {
+			return
+		}
+
+		this.prepullArcanas.forEach(arcanaBuff => {
+			if (arcanaBuff.ability.guid === event.ability.guid
+				&& arcanaBuff.targetID === event.targetID) {
+
+				const cardStateItem: CardState = {..._.last(this.cardStateLog)} as CardState
+				const arcanaAction = getDataBy(ACTIONS, 'id', this.arcanaStatusToPlay(event.ability.guid))
+
+				if (arcanaAction) {
+					const arcanaCastEvent: CastEvent = {
+						ability: {
+							name: arcanaAction.name,
+							guid: arcanaAction.id as number,
+							type: 1,
+							abilityIcon: _.replace(_.replace(arcanaAction.icon, 'https://xivapi.com/i/', ''), '/', '-'),
+						},
+						timestamp: event.timestamp - (STATUSES.THE_BALANCE.duration * 1000),
+						type: 'cast',
+						sourceIsFriendly: true,
+						targetIsFriendly: true,
+						sourceID: event.sourceID,
+						targetID: event.targetID,
+					}
+					cardStateItem.lastEvent = {...arcanaCastEvent}
+					cardStateItem.drawState = DrawnType.NOTHING
+					cardStateItem.sealState = CLEAN_SEAL_STATE
+
+					this.cardStateLog.unshift(cardStateItem)
+					this.pullIndex++
+				}
+
+				return
+			}
+
+		})
+
+	}
+
+	/**
+	 * This will run on removebuff. It will look for the loss of Arcanas Drawn statuses
+	 * 5.0: This was a lot more meaningful from
+	 *
+	 * a) If it can't find any clear reason why the player had lost the buff, let's do a retconsearch to figure out since when they had it
+	 *
+	 * b) If they lost the buff with no link to any timestamp, it could be a /statusoff macro.
+	 *    Creates a new entry as this is technically also a card action.
+	 *
+	 */
+	private offDrawnStatus(event: BuffEvent) {
+
+		if (DRAWN_ARCANA.includes(event.ability.guid)) {
```

Comment:
same story w/ `if`

---

also, `isPaired` can be done with `.some` :meowmorning: